src/Traits/ErrorTrait.php

```diff
@@ -35,67 +48,127 @@ trait ErrorTrait
      * @param string $varName The name of the userinput variable, to be used in the error message.
      * @param string|string[] $classNames The classnames that should be tested against
      * @param mixed $userInput The input that should be tested
+     *
      * @throws TypeError
      */
-    private function assertClass(string $varName, $classNames, $userInput): void
+    private static function assertClass(string $varName, $classNames, $userInput): void
     {
         if (!is_array($classNames)) {
             $classNames = [$classNames];
         }
 
+        if (!self::isClass($classNames, $userInput)) {
+            throw self::typeError($varName, $classNames, $userInput);
+        }
+    }
+
+    /**
+     * Get debug type method stolen from the symfony polyfill
+     *
+     * @see https://github.com/symfony/polyfill/blob/main/src/Php80/Php80.php
+     */
+    public static function getDebugType($value): string
+    {
+        switch (true) {
+            case null === $value: return 'null';
+            case is_bool($value): return 'bool';
+            case is_string($value): return 'string';
+            case is_array($value): return 'array';
+            case is_int($value): return 'int';
+            case is_float($value): return 'float';
+            case is_object($value): break;
```

Comment:
Or put the object stuff inside the case.

Reply:
I don't know for sure if evaluation of the constant-time boolean function in Ghlens' switch statement is less efficient than doing a string comparison. Note that in most php implementations, these functions just read a (few) bit(s) of where the variable is stored, which is in general very fast, and possibly faster than allocating a string containing the type name and checking that.

Consider the following snippet of code:
```php
<?php
namespace A\Very\Ugly\NamespaceName {
 class MyClass {};
}
namespace {

class MyClass {};

function getProposed($value) {
switch(gettype($value)) {
    case 'object':
        $class = get_class($value);
        
        if (strpos($class, '@') !== false) {
                // This is an anonymous class
                $name = get_parent_class($class) ?: 
                    key(class_implements($class)) ?: 
                    'class';

                return "$name@anonymous";
        }

        return $class;
    case 'NULL': return 'null';
    case 'boolean': return 'boolean';
    case 'string': return 'string';
    case 'array': return 'array';
    case 'integer': return 'int';
    case 'double': return 'float';
    case 'resource':
                // For a fair comparison, use get_resource_type here as well.
                if (null === $type = @get_resource_type($value)) {
                    return 'unknown';
                }

                if ('Unknown' === $type) {
                    $type = 'closed';
                }

                return "resource ($type)";
    case 'resource (closed)': return 'resource (closed)';
    case 'unknown type': return 'unknown';
    default:
        throw new LogicException('Invalid or unknown type');
}
}

function getCurrent($value) {

        switch (true) {
            case is_object($value):
                $class = get_class($value);

                if (false === strpos($class, '@')) {
                    return $class;
                }

                return (get_parent_class($class) ?: key(class_implements($class)) ?: 'class').'@anonymous';
            case null === $value: return 'null';
            case is_bool($value): return 'bool';
            case is_string($value): return 'string';
            case is_array($value): return 'array';
            case is_int($value): return 'int';
            case is_float($value): return 'float';
            case $value instanceof __PHP_Incomplete_Class: return '__PHP_Incomplete_Class';
            default:
                if (null === $type = @get_resource_type($value)) {
                    return 'unknown';
                }

                if ('Unknown' === $type) {
                    $type = 'closed';
                }

                return "resource ($type)";
        }
}

file_put_contents('/tmp/testfile.txt', 'testing');
$r = fopen('/tmp/testfile.txt','w');
foreach ([true, null, "string", ["array"], 17, 1.7, new stdClass(), new class(){}, new class() extends MyClass {},  new MyClass(), new A\Very\Ugly\NamespaceName\MyClass(), $r] as $something) {
    $s = hrtime(true);
    for ($i = 0; $i < 10000000; ++$i) {
        $conclusion = getCurrent($something);
    }
    $time1 = hrtime(true) - $s;
    $s = hrtime(true);
    for ($i = 0; $i < 10000000; ++$i) {
        $conclusion = getProposed($something);
    }
    $time2 = hrtime(true) - $s;
    if ($time1 < $time2) {
        $percentage = ((float) $time2-$time1)/((float)$time1);
        echo 'getCurrent';
    } else {
        echo 'getProposed';
        $percentage = ((float) $time1 - $time2)/((float)$time2);
    }
    $percentage = number_format($percentage*100, 2) . '%';
    echo " is $percentage faster for $conclusion.\n";
}
fclose($r);
}

```
I get the following output:

```txtgetCurrent is 14.39% faster for boolean.
getCurrent is 25.41% faster for null.
getCurrent is 4.64% faster for string.
getProposed is 2.98% faster for array.
getProposed is 10.77% faster for int.
getProposed is 18.63% faster for float.
getCurrent is 24.22% faster for stdClass.
getCurrent is 7.98% faster for class@anonymous.
getCurrent is 18.57% faster for MyClass@anonymous.
getCurrent is 24.16% faster for MyClass.
getCurrent is 24.93% faster for A\Very\Ugly\NamespaceName\MyClass.
getProposed is 15.87% faster for resource (stream).
```
So the difference is not very big. One indeed sees that the `switch(true)` statement is slower for the types that are lower on the switch statement (array/int/resource), but it is faster for the statements on the top. I don't think efficiency is much of a concern here, so I think we should focus on readability and reliability of the code.

Reply:
You are right. However, I still don't like the current implementation. We blindly assume that if none of the checks pass, the value _must_ be a resource. What is a future version of PHP adds an additional type?

I also think that just using an `if-else` chain would be more readable than the `switch(true)`, only because it is a much more common pattern.

Reply:
This is an impressive discussion for something I quickly stole from Symfony. I have tried to compromise by using if statements and separating methods, splitting the if statements between objects, scalars and resources.

My two cents:
- If it is good enough for one of the biggest frameworks in the world, it ought to be good enough for us
- If PHP adds a new type, it will be announced a long time beforehand and discussed by thousands of people. In theory, every language construct can be changed or extended, there is no solution to that.
- We can decide to throw an exception if the type was not detected, but that opens another can of worms: do end users need to expect this exception, even though there is no chance it will actually happen? A lot of static analysis tools will yell at them.

Reply:
looks good!